test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperBlockRewriterTest.cs

```diff
@@ -0,0 +1,1095 @@
+                            new MarkupTagHelperBlock(
+                                "input",
+                                selfClosing: false,
+                                attributes: new List<KeyValuePair<string, SyntaxTreeNode>>()
+                                {
+                                    new KeyValuePair<string, SyntaxTreeNode>("bound-required-int", null),
+                                    new KeyValuePair<string, SyntaxTreeNode>("unbound-required", null),
+                                    new KeyValuePair<string, SyntaxTreeNode>("bound-required-string", null),
+                                },
+                                children: new MarkupTagHelperBlock(
+                                    "p",
+                                    selfClosing: false,
+                                    attributes: new List<KeyValuePair<string, SyntaxTreeNode>>()
+                                    {
+                                        new KeyValuePair<string, SyntaxTreeNode>("bound-int", null),
+                                        new KeyValuePair<string, SyntaxTreeNode>("bound-string", null),
+                                    }))),
+                        new[]
+                        {
+                            new RazorError(string.Format(errorFormatNoCloseAngle, "input"), SourceLocation.Zero),
+                            new RazorError(string.Format(errorFormatUnclosed, "input"), SourceLocation.Zero),
+                            new RazorError(
+                                string.Format(errorFormatNoValue, "bound-required-int", "input", intType), 7, 0, 7, 18),
+                            new RazorError(
+                                string.Format(errorFormatNoValue, "bound-required-string", "input", stringType),
+                                absoluteIndex: 43,
+                                lineIndex: 0,
+                                columnIndex: 43,
+                                length: 21),
+                            new RazorError(string.Format(errorFormatNoCloseAngle, "p"), 64, 0, 64),
+                            new RazorError(string.Format(errorFormatUnclosed, "p"), 64, 0, 64),
+                            new RazorError(string.Format(errorFormatNoValue, "bound-int", "p", intType), 67, 0, 67, 9),
+                            new RazorError(
+                                string.Format(errorFormatNoValue, "bound-string", "p", stringType),
+                                absoluteIndex: 77,
+                                lineIndex: 0,
+                                columnIndex: 77,
+                                length: 12),
+                        }
+                    },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(MinimizedAttributeData_Document))]
+        [MemberData(nameof(MinimizedAttributeData_CSharpBlock))]
+        [MemberData(nameof(MinimizedAttributeData_PartialTags))]
```

Comment:
The Razor parser is a recursive-descent parser with a fair number of special cases, I agree.  However it isn't so fragile that changes in the low-level attribute parsing merit so many unreadable new cases.  For example you've added a single product code clause in the condition about minimized-null-or-empty values for bound non-`string` attributes, not enough to justify testing minimized `string` attributes, minimized unbound attributes, and so on.

Bottom line, it isn't possible to see what matters in this test class.